File: line2d.java
/*
 * VAST, a scalable peer-to-peer network for virtual environments
 * Copyright (C) 2006 Shun-Yun Hu (syhu@yahoo.com)
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

package vast;

public class line2d
{
    public double  a,b,c;
    public int     bisectingID[] = new int[2];    // NOTE: we need to change this from storing node index to node id
    public int     vertexIndex[] = new int[2];
    public segment seg;
    
    public line2d (double x1, double y1, double x2 , double y2)
    {
        seg = new segment (x1, y1, x2, y2);
        
        if (y1 == y2) 
        {   
            a = 0; b = 1; c = y1;
        }
        else if (x1 == x2) 
        {
            a = 1; b = 0; c = x1;
        }
        else
        {
            double dx = x1 - x2; 
            double dy = y1 - y2;
            double m = dx / dy;
            a = -1 * m;
            b = 1;
            c = a*x1 + b*y1;
        }
    }

    public line2d () 
    {
        a = b = c = 0.0;
        seg = new segment (0, 0, 0, 0);
        
        vertexIndex[0] = -1;
        vertexIndex[1] = -1;
    }

    public line2d (double A, double B, double C)
    {
        a = A;
        b = B;
        c = C;
        seg = new segment (0, 0, 0, 0);
        
        vertexIndex[0] = -1;
        vertexIndex[1] = -1;
    }

    // BUG: possibily not storing p properly after calculation
    boolean intersection (line2d l, point2d p)
    {
        //The polynomial judgement
        double D = (a * l.b) - (b * l.a);
        if (D == 0) 
        {
            p.x = 0;
            p.y = 0;
            return false;
        }
        else
        {
            p.x=(c*l.b-b*l.c)/D/1.0;
            p.y=(a*l.c-c*l.a)/D/1.0;
            return true;
        }   
    }

    double dist (point2d p) 
    {        
        /*
        double u;

        // u = ((x3-x1)(x2-x1) + (y3-y1)(y2-y1)) / (p2.dist(p1)^2)
        u = ((p.x - seg.p1.x) * (seg.p2.x - seg.p1.x) + 
             (p.y - seg.p1.y) * (seg.p2.y - seg.p1.y)) /
             (seg.p1.dist (seg.p2))

        double x = seg.p1.x + u * (seg.p2.x - seg.p1.x);
        double y = seg.p1.y + u * (seg.p2.y - seg.p1.y);

        return p.dist (point2d (x, y));
        */

        return Math.abs (a * p.x + b * p.y + c) / Math.sqrt (Math.pow (a, 2) + Math.pow (b, 2));
    }
};

File: lib/original/typedef/line2d.js
// Include the necessary modules if not defined

var point2d = point2d || require( "./point2d.js" );
var segment = segment || require( "./segment.js" ); 

//function line2d (x1, y1, x2, y2)
function line2d(a, b, c)
{
    // public
    this.a = (typeof a == 'undefined' ? 0 : a);                 // double
    this.b = (typeof b == 'undefined' ? 0 : b);
    this.c = (typeof c == 'undefined' ? 0 : c);                 
    this.seg = new segment();   // segment
    
    //var  bisecting = new Array();         // int
    this.bisectingID = [];          // int[2]
    this.vertexIndex = [];          // int[2]
    
	this.vertexIndex[0] = -1;
	this.vertexIndex[1] = -1;
    
    // method to set the two endpoints for a line
    this.set = function (x1, y1, x2, y2) {
    
        // store coordinates
        seg.p1 = new point2d(x1, y1);
        seg.p2 = new point2d(x2, y2);
    
		if (y1 == y2) {	
			a = 0; b = 1; c = y1;
		}
		else if (x1 == x2) {
			a = 1; b = 0; c = x1;
		}
		else {
			var dx = x1 - x2;       // double
			var dy = y1 - y2;       // double
			var m = dx / dy;        // double
            
			a = -1 * m;
			b = 1;
			c = a*x1 + b*y1;
		}    
    }
        
    // line2d l, point2d p
    // The polynomial judgement!
	this.intersection = function (l , p) {
		
		var D = (a * l.b) - (b * l.a);
        
		if (D == 0) {
			p.x = 0;
			p.y = 0;
			return false;
		}
		else {
			p.x=(c*l.b - b*l.c) / D / 1.0;
			p.y=(a*l.c - c*l.a) / D / 1.0;
            return true;
		}	
	}

    this.distance = function (p) {        
        return Math.abs (a * p.x + b * p.y + c) / Math.sqrt (Math.pow (a, 2) + Math.pow (b, 2));
    }
};

if (typeof module !== 'undefined')
	module.exports = line2d;
